homeassistant/components/nanoleaf/config_flow.py

```diff
@@ -0,0 +1,205 @@
+        name = discovery_info["name"].replace(f".{discovery_info['type']}", "")
+        await self.async_set_unique_id(name)
+        self._abort_if_unique_id_configured({CONF_HOST: host})
+        self.nanoleaf = Nanoleaf(host)
+
+        # Import from discovery integration
+        self.device_id = discovery_info["properties"]["id"]
+        self.discovery_conf = cast(
+            dict,
+            await self.hass.async_add_executor_job(
+                load_json, self.hass.config.path(CONFIG_FILE)
+            ),
+        )
+        self.nanoleaf.token = self.discovery_conf.get(self.device_id, {}).get(
+            "token",  # >= 2021.4
+            self.discovery_conf.get(host, {}).get("token"),  # < 2021.4
+        )
+        if self.nanoleaf.token is not None:
+            self.context["source"] = config_entries.SOURCE_INTEGRATION_DISCOVERY
+            _LOGGER.warning(
+                "Importing Nanoleaf %s from the discovery integration", name
+            )
+            return await self.async_setup_finish()
+
+        self.context["title_placeholders"] = {"name": name}
+        return await self.async_step_link()
+
+    async def async_step_link(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Handle Nanoleaf link step."""
+        if user_input is None:
+            return self.async_show_form(step_id="link")
+
+        try:
+            await self.hass.async_add_executor_job(self.nanoleaf.authorize)
+        except NotAuthorizingNewTokens:
+            return self.async_show_form(
+                step_id="link", errors={"base": "not_allowing_new_tokens"}
+            )
+        except Unavailable:
+            return self.async_abort(reason="cannot_connect")
+        except Exception:  # pylint: disable=broad-except
+            _LOGGER.exception("Unknown error authorizing Nanoleaf")
+            return self.async_show_form(step_id="link", errors={"base": "unknown"})
+        return await self.async_setup_finish()
+
+    async def async_step_import(self, config: dict[str, Any]) -> FlowResult:
+        """Handle Nanoleaf configuration import."""
+        self._async_abort_entries_match({CONF_HOST: config[CONF_HOST]})
+        _LOGGER.debug(
+            "Importing Nanoleaf on %s from your configuration.yaml", config[CONF_HOST]
+        )
+        self.nanoleaf = Nanoleaf(config[CONF_HOST])
+        self.nanoleaf.token = config[CONF_TOKEN]
+        return await self.async_setup_finish()
+
+    async def async_setup_finish(self) -> FlowResult:
+        """Finish Nanoleaf config flow."""
+        try:
+            info = await self.hass.async_add_executor_job(
+                pynanoleaf_get_info, self.nanoleaf
+            )
+        except Unavailable:
+            return self.async_abort(reason="cannot_connect")
+        except InvalidToken:
+            return self.async_show_form(
+                step_id="link", errors={"base": "invalid_token"}
+            )
```

Comment:
This shouldn't happen in a user flow because we just got the token.
So it only happens with an import where the token is invalid. Should it abort the flow here too?

Reply:
Yes, abort on invalid imports seems fine to me.